### PR TITLE
Adds nodes usage API to monitor usages of actions

### PR DESCRIPTION
**Pull request by colings86**
The nodes usage API has 2 main endpoints

/_nodes/usage and /_nodes/{nodeIds}/usage return the usage statistics
for all nodes and the specified node(s) respectively.

At the moment only one type of usage statistics is available, the REST
actions usage. This records the number of times each REST action class is
called and when the nodes usage api is called will return a map of rest
action class name to long representing the number of times each of the action
classes has been called.

Still to do:

* [x] Create usage service to store usage statistics
* [x] Record usage in REST layer
* [x] Add Transport Actions
* [x] Add REST Actions
* [x] Tests
* [x] Documentation
* [ ] Check for references to statistics and replace with "feature usage"

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/rest/BaseRestHandler.java [hunk 1]
```diff
@@ -54,10 +55,18 @@
...
+
     protected BaseRestHandler(Settings settings) {
         super(settings);
     }
 
+    public long getUsageCount() {
+        return usageCount.sum();
+    }
+
+    public abstract String getName();
```
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/rest/BaseRestHandler.java]: can you add javadocs about what the name should be

File: core/src/main/java/org/elasticsearch/rest/BaseRestHandler.java [hunk 2]
```diff
@@ -54,10 +55,18 @@
     public static final Setting<Boolean> MULTI_ALLOW_EXPLICIT_INDEX =
         Setting.boolSetting("rest.action.multi.allow_explicit_index", true, Property.NodeScope);
 
+    private final LongAdder usageCount = new LongAdder();
+
     protected BaseRestHandler(Settings settings) {
         super(settings);
     }
 
+    public long getUsageCount() {
```
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/rest/BaseRestHandler.java]: maybe make this final?

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesUsageAction.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        String[] nodesIds = Strings.splitStringByCommaToArray(request.param("nodeId"));
+        Set<String> metrics = Strings.splitStringByCommaToSet(request.param("metric", "_all"));
+
+        NodesUsageRequest nodesUsageRequest = new NodesUsageRequest(nodesIds);
+        nodesUsageRequest.timeout(request.param("timeout"));
+
+        if (metrics.size() == 1 && metrics.contains("_all")) {
```
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesUsageAction.java]: should we handle the case where `_all` is sent with `rest_actions`?

File: core/src/main/java/org/elasticsearch/usage/UsageService.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A service to monitor usage of Elasticsearch features.
+ */
+public class UsageService extends AbstractComponent {
+
+    private final List<BaseRestHandler> handlers;
+    private long sinceTime;
```
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/usage/UsageService.java]: this can be final?

**Review by jaymode — CHANGES_REQUESTED** [review]
left a few minor comments

**Comment by colings86** [comment]
@jaymode thanks for reviewing, I updated the PR to address your comments

**Review by jaymode — APPROVED** [review]
LGTM

**Pull request merged**